src/software/embedded/robot_diagnostics_cli/robot_diagnostics_cli.py

```
@@ -0,0 +1,517 @@
+        """Forward PrimitiveSet protos from diagnostics to the robots.
+
+        Updates/polls the Estop state
+
+        If the emergency stop is tripped, the PrimitiveSet will not be sent so
+        that the robots timeout and stop.
+        """
+        self.__update_estop_state()
+        if self.should_send_stop:
+            raise KeyboardInterrupt
+        else:
+            self.__send_primitive_set(diagnostics_primitive)
+
+    def catch_interrupt_exception(exit_code=1):
+        """Decorator for handling keyboard exceptions and safely clearing cached primitives"""
+
+        def decorator(func):
+            @wraps(func)
+            def wrapper(self, *args, **kwargs):
+                try:
+                    self.__run_primitive_set(Primitive(stop=StopPrimitive()))
+                    return func(self, *args, **kwargs)
+                except KeyboardInterrupt:
+                    print("E-Stop Activated: Stopped Primitive Send")
+                    self.__run_primitive_set(Primitive(stop=StopPrimitive()))
+                    raise Typer.Exit(code=exit_code)
+                except Exception as e:
+                    self.__run_primitive_set(Primitive(stop=StopPrimitive()))
+                    print(f"Unknown Exception: {e}")
+                    raise Typer.Exit(code=exit_code)
+                finally:
+                    self.__run_primitive_set(Primitive(stop=StopPrimitive()))
+            return wrapper
+
+        return decorator
+
+    def __update_estop_state(self) -> None:
+        """Updates the current estop status proto if estop is not disabled
+        Always in physical estop mode, uses the physical estop value
+        """
+        if self.estop_mode == EstopMode.DISABLE_ESTOP:
+            self.should_send_stop = False
+        elif self.estop_mode == EstopMode.PHYSICAL_ESTOP:
+            self.should_send_stop = not self.estop_reader.isEstopPlay()
+        else:
+            self.should_send_stop = True
```

you should raise an exception here

---

Would that be necessarily the best? I wish to keep the number of exceptions floating around to a minimum since I feel it makes the logic structure hard to follow. Also, `self.should_send_stop` is in itself what raises the exception in the first place. 

The reason this update check is isolated from the `run_primitive` func is so that this check can be performed/updated lonesome at the startup, before a command is entered yet.